scheds/rust/scx_rustland/src/bpf/main.bpf.c

```diff
@@ -307,30 +346,58 @@ static void dispatch_task(struct task_struct *p, u64 dsq_id, u64 enq_flags)
 		/*
 		 * Dispatch a task to a specific per-CPU DSQ if the target CPU
 		 * can be used (according to the cpumask), otherwise redirect
-		 * the task to the CPU assigned by the built-in idle selection
+		 * the task to the first CPU available, using the shared DSQ
 		 * logic.
 		 *
-		 * If also this CPU is not usable (because cpumask has changed
-		 * in the meantime), we can simply ignore the task, as it has
-		 * been dequeued and re-enqueued, so it will pick a valid CPU.
+		 * This can happen if the user-space scheduler dispatches the
+		 * task to an invalid CPU, the redirection to the shared DSQ
+		 * allows to prevent potential stalls in the scheduler.
+		 *
+		 * If the cpumask is not valid anymore (determined by the
+		 * cpumask_cnt generation counter) we can simply cancel the
+		 * dispatch event, since the task will be re-enqueued by the
+		 * core sched-ext code, potentially selecting a different cpu
+		 * and a different cpumask.
 		 */
 		scx_bpf_dispatch(p, dsq_id, slice, enq_flags);
 
+		tctx = lookup_task_ctx(p);
+		if (!tctx)
+			return;
+
+		/* Read current cpumask generation counter */
+		curr_cpumask_cnt = __sync_fetch_and_add(&tctx->cpumask_cnt, 0);
```

Comment:
nit: I don't think __sync op is necessary here, `curr_cpumask_cnt = tctx->cpumask` should be sufficient. All the involved operations are already fully synchronized.